prow/resultstore/uploader.go

```diff
@@ -58,9 +62,54 @@ func (u *Uploader) Upload(ctx context.Context, log *logrus.Entry, p *Payload) er
 	w.WriteTarget(ctx, p.OverallTarget())
 	w.WriteConfiguredTarget(ctx, p.ConfiguredTarget())
 	w.WriteAction(ctx, p.OverallAction())
-	// TODO: When writer.New() is emended to resume uploads (see TODO there),
-	// return non-permanent errors from Finalize(). Doing this now is futile,
-	// since attempting to upload an existing invocation is a permanent error.
-	w.Finalize(ctx)
+
+	if err := w.Finalize(ctx); err != nil {
+		if writer.PermanentError(err) {
+			return nil
+		}
+		return err
+	}
 	return nil
 }
+
+type tokener struct {
+	mu   sync.Mutex
+	seed []byte
+}
+
+func (t *tokener) For(id string) string {
+	t.mu.Lock()
+	seed := t.seed
+	t.mu.Unlock()
+
+	digest := sha256.New()
+	digest.Write(seed)
+	digest.Write([]byte(id))
+	r := bytes.NewReader(digest.Sum(nil))
+	// error cannot occur since we read from a 256-bit digest.
+	u, _ := uuid.NewRandomFromReader(r)
+	return u.String()
+}
+
+func (t *tokener) Reseed(seed string) {
+	digest := sha256.New()
+	digest.Write([]byte(seed))
+
+	t.mu.Lock()
+	t.seed = digest.Sum(nil)
+	t.mu.Unlock()
+}
+
+var authToken *tokener
+
+func init() {
+	authToken = &tokener{}
+	SeedAuthToken("Avast ye, Matey!")
+}
+
+// SeedAuthToken updates the seed for computing AuthenticationToken values
+// for ResultStore uploads. Using an unchanging secret value best prevents
+// unintended writers from interfering with in-progress uploads.
+func SeedAuthToken(seed string) {
+	authToken.Reseed(seed)
+}
```

Comment:
Nit: Why not just hardcode the string inside "Reseed" if we're never going to use a different string? I don't understand the purpose of the extra layer of indirection to let Reseed be parameterized.